plugins/source_scheduler_dfs.go

```diff
@@ -0,0 +1,192 @@
+			}
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				defer p.tableSem.Release(1)
+				// not checking for error here as nothing much todo.
+				// the error is logged and this happens when context is cancelled
+				p.resolveTableDfs(ctx, table, client, nil, resolvedResources)
+			}()
+		}
+	}
+	wg.Wait()
+}
+
+func (p *SourcePlugin) resolveTableDfs(ctx context.Context, table *schema.Table, client schema.ClientMeta, parent *schema.Resource, resolvedResources chan<- *schema.Resource) {
+	clientName := client.Name()
+	logger := p.logger.With().Str("table", table.Name).Str("client", clientName).Logger()
+	logger.Info().Msg("table resolver started")
+	tableMetrics := p.metrics.TableClient[table.Name][clientName]
+	res := make(chan interface{})
+	go func() {
+		defer func() {
+			if err := recover(); err != nil {
+				stack := fmt.Sprintf("%s\n%s", err, string(debug.Stack()))
+				sentry.WithScope(func(scope *sentry.Scope) {
+					scope.SetTag("table", table.Name)
+					sentry.CurrentHub().CaptureMessage(stack)
+				})
+				p.logger.Error().Interface("error", err).Str("stack", stack).Msg("table resolver finished with panic")
+				atomic.AddUint64(&tableMetrics.Panics, 1)
+			}
+			close(res)
+		}()
+		if err := table.Resolver(ctx, client, parent, res); err != nil {
+			logger.Error().Err(err).Msg("table resolver finished with error")
+			atomic.AddUint64(&tableMetrics.Errors, 1)
+			return
+		}
+	}()
+
+	for r := range res {
+		p.resolveResourcesDfs(ctx, table, client, parent, r, resolvedResources)
+	}
+
+	// we don't need any waitgroups here because we are waiting for the channel to close
+	logger.Info().Uint64("resources", tableMetrics.Resources).Uint64("errors", tableMetrics.Errors).Msg("fetch table finished")
+}
+
+func (p *SourcePlugin) resolveResourcesDfs(ctx context.Context, table *schema.Table, client schema.ClientMeta, parent *schema.Resource, resources interface{}, resolvedResources chan<- *schema.Resource) {
+	resourcesSlice := helpers.InterfaceSlice(resources)
+	if len(resourcesSlice) == 0 {
+		return
+	}
+	resourcesChan := make(chan *schema.Resource, len(resourcesSlice))
+	go func() {
+		defer close(resourcesChan)
+		var wg sync.WaitGroup
+		for i := range resourcesSlice {
+			i := i
+			if err := p.resourceSem.Acquire(ctx, 1); err != nil {
+				p.logger.Warn().Err(err).Msg("failed to acquire semaphore. context cancelled")
+				wg.Wait()
+				// we have to continue emptying the channel to exit gracefully
+				return
+			}
+			wg.Add(1)
+			go func() {
+				defer p.resourceSem.Release(1)
+				defer wg.Done()
+				//nolint:all
+				resolvedResource := p.resolveResource(ctx, table, client, parent, resourcesSlice[i])
+				if resolvedResource == nil {
+					return
+				}
+				resourcesChan <- resolvedResource
+			}()
+		}
+		wg.Wait()
+	}()
+
+	for resource := range resourcesChan {
+		resolvedResources <- resource
+		if resource.Table.Relations == nil {
+			continue
+		}
+		for _, relation := range resource.Table.Relations {
```

Comment:
We should consider parallelizing this part as well (future PR), as I think it would benefit us more at the moment.
I don't think many plugins use `PreResourceResolver` for list/detail pattern yet

Reply:
Yeah I think there is much benchmarking to do but for now I'd keep it simple (it is basically the same scheduling as we had before just without the deadlock)